Add a section on the purpose of DCV (#160)
ietf-wg-dnsop/draft-ietf-dnsop-domain-verification-techniques: draft-ietf-dnsop-domain-verification-techniques.md

* `Random Token`: a random value that uniquely identifies the DNS domain control validation challenge, defined in {{random-token}}.

# Purpose of Domain Control Validation {#purpose}

Domain Control Validation is a challenge-response procedure that allows the User to prove operational control of the contents of a domain to an Application Service Provider. This proof applies only to the moment in time when the record is published. This procedure can be appropriate when the Application Service Provider is about to take a time-bounded action related to this domain name.

> Domain Control Validation is not suitable for ongoing authorization. Any ongoing authorization using DNS SHOULD be performed with a separate DNS record. For example, in the ACME protocol, issuance of a time-limited certificate can use Domain Control Validation with an ephemeral TXT record via the DNS-01 challenge mechanism ({{RFC8555, Section 8.4}}), but ongoing authorization of certificate authorities uses a persistent CAA record {{?RFC8569}}.

Reviewer: I don't understand this part. Why is it not suitable?

Whether the "still valid" notion appears in a DNS-01 or CAA record doesn't really matter ? In the ACME case, it is just that you can convey a change (eg only allow different registar) that is different from a challence sent once for a single registrar.

Reviewer: I agree that new challanges would be better to proof this, but I don't see why a split into a different record would matter

Reviewer 2: Maybe the point is that domain control validation (proving that you have control over a domain) is a distinct action than the domain controller authorizing some other service to do something for it, so lumping these into the same record may not make sense?

The authorization action is typically ongoing, whereas the DCV action is point in time. Ongoing authorization uses some constant piece of rdata, whereas DCV typically uses a nonce.

Reviewer: I understand, but I see no technical reason these cannot be combined, provided that one feels ok with a permanent entry in the zone versus a freshness proof, which in most scenarios is infeasable or not done (eg I dont need to do a monthly challenge to keep using pardot :P

Author: > I see no technical reason these cannot be combined

It can certainly be made to work, but this is a BCP, so the question is: is that the _best_ way?

I think the best way is to have a very clear separation between "ephemeral proof of control" and "persistent authorization".  Some reasons:

* It avoids errors of reasoning where someone checks the validity of the record and concludes that it demonstrates control, when that is no longer true.
* It encourages use of meaningful zone content for authorization, instead of opaque identifiers.  This aids debugging and removes the need for a database to track the meaning of various high-entropy blobs.
* It avoids accidental outages when DCV records are removed after the challenge is completed (because they _look_ just like the ephemeral challenge records that are usually safe to remove).

Reviewer 3: I also disagree with the premise of this (ie, that DCV is necessarily time-bounded). There are critical use-cases where recurring re-validation is needed.  As an example, with ACME a common use-case is for the _acme-challenge.www.example.com to be a CNAME to an intermediate (eg, on a CDN) that the intermediate uses on a recurring basis to re-validate new certificates.  The ongoing presence of the _acme-challenge.www.example.com is not time-bounded and has meaning that the intermediate can continue to revalidate new certificates.  There are likely similar use-cases for this.  The whole DV cert ecosystem relies heavily on this.

I also see CAA as being orthogonal to DCV.  We'd talked about this some but decided to leave it out other than a brief mention.  DCV provides explicit proof of control (whether ephemeral or persistent) while CAA is constraining the allowed set of control. What is allowed is the intersection between these two.  This pattern may be useful elsewhere (eg, we may want a separate draft similar to this one that generalizes the DNS-based constraints that CAA does (but which fixes some of the CAA issues), but this is separate than CAA providing a persistent authorization.  Including "LetsEncrypt,Digicert" in a CAA record doesn't authorize to LetsEncrypt or Digicert but rather says that anything outside of those two is prohibited.

(Separately there are some highly problematic protocol design issues with CAA from a DNS perspective, but that is not relavant here.)

Author: > The ongoing presence of the _acme-challenge.www.example.com is not time-bounded

The CNAME is not a DCV record.  It is an ongoing authorization, and is a separate record from the DCV record as recommended.  That's why this PR adjusts the example CNAME to contain no entropy, to clarify the distinction.  (DCV records necessarily contain high entropy.)

Reviewer 3: Why is the presence of entropy correlated with the time-duration?  While I see your points, our experience after trying some variations here is that a high-entropy blog tracked in a DB but which is otherwise meaningless is safer, simpler, and easier to reason about than a target name which either leaks information or is derived somehow.

Where we may get stuck here is the tension between information in the domain being public and wanting to keep book-keeping information for future operational purposes.  If there was a way to keep structured comment information which was never returned in queries then that could help here, but just putting potentially sensitive information into the zone to help keep track of things doesn't help.

I'm also not sure I see why these high-entropy tokens by themselves are disqualifying for long-term validation.

Reviewer 4: This thread is getting a bit long to usefully reply to, but I will try to make a few points that seem to tie to unresolved topics:
- the high-entropy token serves to bind the validation to the account at the provider (the provider only issues the token to the specific account) but it is a single-use token for that purpose (once it is published in the DNS everybody knows it and it is no longer useful for binding the domain validation to the account on an ongoing basis).
- I agree that CAA and the CNAME to an intermediate are conveying ongoing authorization information rather than domain validation information directly.  We still want to use a random value for the CNAME target to get account binding for the initial iteration (see previous bullet point), and that points to a scenario where there are some intrinsically non-separable behaviors that combine the one-time/ephemeral and ongoing properties of a single DNS record.
- I think the (proposed) text as-is is confusing/unhelpful, since it is essentially an appeal to authority ("ACME does it this way, so that must be good").  It would be better to have more discussion of how these are related but distinct topics and why we might choose to handle them distinctly.

Reading the above, I'm coming around to a stance that there is both a time-bounded and a potentially longer-lived aspect to DCV -- the time-bounded part is the binding of (account at provider) to (domain validation instance), but that a given validation record can remain in place to indicate an ongoing desire (or authorization if you want) to continue doing the operation that was initiated with the initial validation.  I would be ok recommending a different record for ongoing authorization that is entirely separate from any domain validation (which CAA is), but I don't think that a blanket statement forbidding combining domain validation and ongoing authorization is workable.

* `Random Token`: a random value that uniquely identifies the DNS domain control validation challenge, defined in {{random-token}}.

# Purpose of Domain Control Validation {#purpose}

> Domain Control Validation (DCV) is a challenge-response protocol that allows one party (the User) to prove operational control of the contents of a domain to an Application Service Provider (ASP).  This proof applies at a single point in time; it does not provide evidence of continuing control prior to the challenge or after the record is published.

Reviewer 3: ```suggestion
Domain Control Validation (DCV) is a pattern for challenge-response protocols that allows one party (the User) to prove operational control of the contents of a domain to an Application Service Provider (ASP).  This proof applies at a single point in time; it does not provide evidence of continuing control prior to the challenge or after the record is published.
```

Author: Changed to "procedure"

> Domain Control Validation (DCV) is a challenge-response protocol that allows one party (the User) to prove operational control of the contents of a domain to an Application Service Provider (ASP).  This proof applies at a single point in time; it does not provide evidence of continuing control prior to the challenge or after the record is published.

Reviewer 3: Should we move the "(ASP)" up to the definitions above where we define Application Service Provider, or perhaps just expand ASP throughout?

Reviewer 5: My suggestion is add the (ASP) to the definition, but for now do not use it (so we expand it in this pull request). 
My logic is have it in place so it can be used - rfc editor or someone may think it is a good idea

Author: I've removed the acronyms now that this is much shorter.  I'll let the authors decide how to proceed.

DCV is appropriate when the ASP is about to take a brief, time-bounded action that is potentially dangerous but not more harmful than the damage that could be done by malicious modification to the DNS zone.

Appropriate uses of DCV:

* Issuing a time-bounded TLS certificate via an ACME "dns-01" challenge.
  - Control of the DNS zone would be sufficient to acquire a certificate anyway via the ACME "http-01" challenge.
* Allowing a user to claim ownership of a domain name in a third-party system for a limited period of time.

>   - This creates a risk of impersonation, but the User has already demonstrated their ability to impersonate this name.

Reviewer 5: expand ASP in therse cases

also

s/provision resources on the domain name/provision resources of the domain name/

>   - This creates a risk of impersonation, but the User has already demonstrated their ability to impersonate this name.

Author: In keeping with the draft's move toward brevity, I've removed these various examples.

* Sending a large batch of TLS authenticated requests to a server using this name.

>   - This creates a risk of outage due to overload, but the User could create an outage anyway by deleting important records.

Author: Yes.  (But I've removed this section.)

Inappropriate uses of DCV:

* Allowing the User to claim ownership of the domain name indefinitely in a third-party system.

>   - The User could be an attacker who has only transient control of the domain name, or the domain name could be transferred to a new owner.

Reviewer 3: What is the practical risk of this when combined with another delegation such as a CNAME to the ASP?  The CNAME is what delegates to the ASP and the DCV entry is what demonstrates which of the ASP's customers/accounts has control.  How practical is this "stale control" risk?  Would leaving the DCV entry in-place while still using the service suffice (which means we'd want to update some of the "removal" guidance?  The assumption that a new future owner of a zone would remove or not create the DCV entry and if they wanted to use the ASP they'd revalidate there (so perhaps some guidance to the effect of this use-case). At some point the operational burden of redoing DCV at some frequency will overweigh its value and it won't get implemented at all.

Author: Delegated DCV is a curious construct.  It seems to fuse persistent authorization and ephemeral control verification.  This is puzzling.

On closer inspection, the version of Delegated DCV proposed here seems to have a notable deficiency: it cannot mitigate dangling subdomain attacks after the domain is live.  It also uses a random token where randomness is not needed, which I think obscures the security intent.

Reviewer 3: If we revisit the premise of whether DCV provides persistent authorization then perhaps this is less of a confusing construct?  I'm not sure they are distinct -- there are use-cases where both may want to be persistent and perhaps instead we should focus on what is needed to make that (unchanging) persistence safer.

* Sending a large batch of unauthenticated requests to the IP addresses indicated for this domain name.
  - The User could have placed the IP addresses of an unaffiliated victim server on the domain name.  Sending a large volume of requests without authenticating the server would produce an attack that the User would not otherwise have been able to accomplish.

> If the ASP requires ongoing proof of control, DCV can be used repeatedly.  However, for the sake of reliability and efficiency, it may be preferable to use a DNS record that persistently authorizes a specified external entity to act on its behalf.  This is consistent with the approach used by most ordinary DNS records, including AAAA and MX.

Reviewer 3: On the "consistent with the approach", this skips past the vulnerabilities in that approach which has lead to DCV, such as "dangling CNAME/MX/DNS/NS record" vulnerabilites.  We may want to touch on this here (as I alluded to above).  Perhaps:

```suggestion
If the ASP requires ongoing proof of control, DCV can be used repeatedly.  However, for the sake of reliability and efficiency, it may be preferable to use a DNS record that persistently authorizes a specified external entity to act on its behalf.  This is consistent with the approach used by most ordinary DNS records, including AAAA, MX, CNAME, and NS.  The reason why those records are not sufficient on their own is that they delegate control of the domain name to an ASP, but they do not indicate to the ASP which of its Users has permissions to provision or otherwise configure the domain name.
```

Author: Agreed, but I've removed this example for brevity.

> 

Reviewer 3: This section might be a good opportunity to address the Discuss from @kaduk on "Recommendation for random token".

```suggestion
## Purpose of Domain Control Validation records  {#dcv-record-purpose}

Within DCV, DNS records containing tokens are used to demonstrate that a User has control over a domain name.  What matters here is:
* collision-free uniqueness -- preventing collisions with pre-existing or guessable records that might enable an attacker to confuse the ASP into believing the attacker has control 
* causal flow -- enabling the ASP to demonstrate a causal relationship from them providing a token to a User for a specific name in the specific context to the token appearing in a DCV record
* conveyance of the intended action being confirmed in a way that a DNS Administrator can validate, including what is being validated as well as the scope of the validation
```

# Delegated Domain Control Validation {#delegated}

Delegated domain control validation lets a User delegate the domain control validation process for their domain to an Intermediary without granting the Intermediary the ability to make changes to their domain or zone configuration.  It is a variation of TXT record validation ({{txt-record}}) that inserts a CNAME record prior to the TXT record.

The Intermediary gives the User a CNAME record to add for the domain and Application Service Provider being validated that points to the Intermediary's domain, where the actual validation TXT record is placed.  The Intermediary's domain MUST be unique to the User, domain, and Application Service Provider. For example, if the User is Example Inc, then the CNAME record could be:

    _service-challenge.example.com.  IN   CNAME  example.com.service.example-inc.dcv.intermediary.example.

The Intermediary then adds the actual Validation Record in a domain they control:

    example.com.service.example-inc.dcv.intermediary.example.  IN   TXT  "<provider-random-token>"

> 

Reviewer 4: The intermediary needs to ensure that the token is bound to the customer in question, but that binding does not need to be exposed to everyone, yes.

Such a setup is especially useful when the Application Service Provider wants to periodically re-issue the challenge with a new provider random token. CNAMEs allow automating the renewal process by letting the Intermediary place the random token in their DNS zone instead of needing continuous write access to the User's DNS.

> During provisioning and renewal, or whenever ownership of a domain changes, there is the risk that an attacker could leverage a "dangling CNAME" to perform a "subdomain takeover" attack ({{SUBDOMAIN-TAKEOVER}}).  Using a subdomain tied to the User prevents these attacks.  This can be accomplished by choosing a name that includes the User's account name (as shown above) or another unique token linked to the User (as in {{random-token}}).

Reviewer 4: (continuing the theme from above) I would prefer to not suggest using a name that includes the User's account name, since that is essentially some information disclosure that does not serve much purpose.  It maybe gives a little nudge to the provider to perform the User-account binding of the token but there's not any actual in-protocol mechanism to enforce that the provider does the binding, so I think our guidance here would be approximately as strong of a mechanism to get that binding.
